File: Leetcode/cpp/medium/1888.cpp
/*
 * @lc app=leetcode id=1888 lang=cpp
 *
 * [1888] Minimum Number of Flips to Make the Binary String Alternating
 */


// @lcpr-template-start
#include <bits/stdc++.h>
using namespace std;
// @lcpr-template-end
// @lc code=start
class Solution {
public:
    int minFlips(string s) {
        int n = s.size();
        int ans = n;
        int cnt = 0;
        for (int r = 0; r < n * 2 - 1; r++) {
            if ((s[r % n] ^ r) & 1) cnt++;
            if (r >= n - 1) {
                ans = min({ans, cnt, n - cnt});
                int left = r - n + 1;
                if ((s[left] ^ left) & 1) cnt--;
            }
        }
        return ans;
    }
};
// @lc code=end

File: Leetcode/rust/medium/1888.rs
/*
 * @lc app=leetcode id=1888 lang=rust
 *
 * [1888] Minimum Number of Flips to Make the Binary String Alternating
 */


// @lcpr-template-start
struct Solution;
// @lcpr-template-end
// @lc code=start
use std::cmp::min;

impl Solution {
    pub fn min_flips(s: String) -> i32 {
        let s = s.as_bytes();
        let n = s.len();
        let mut ans = n;
        let mut cnt = 0;
        for r in 0..(n * 2 - 1) {
            if (s[r % n] as usize ^ r) & 1 != 0 {
                cnt += 1;
            }
            if r >= n - 1 {
                ans = min(ans, min(cnt, n - cnt));
                let left = r - n + 1;
                if (s[left % n] as usize ^ left) & 1 != 0 {
                    cnt -= 1;
                }
            }
        }
        ans as i32
    }
}
// @lc code=end

